### PR TITLE
Rename fields for VSphereCPIConfig CR.

### PR DESCRIPTION
**Pull request by HanFa**
### What this PR does / why we need it

Addresses part of the comments in the VSphereCPIConfig GVK design document. Rename and organize some fields in the VSphereCPIConfig CR.

* rename `spec.server` to `spec.vcenterAPIEndpoint`
* rename `spec.nsxt.host` to `spec.nsxt.apiHost`
* rename `spec.insecureFlag` to `spec.insecure`
* rename `spec.nsxt.insecureFlag` to `spec.nsxt.insecure`
* rename `spec.routes` to `spec.route`
* un-expose the field `spec.routes.clusterCIDR`
* remove the redundant field `spec.nsxt.secretName` and `spec.nsxt.secret`
* group proxy related fields under `spec.proxy`
* group network config related fields under `spec.vmnetwork`
* change all optional fields to be pointers

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

Integration test for this CR

### Release note

```release-note
package-based-lcm: rename fields for the VSphereCPIConfig CR.
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: apis/cpi/v1alpha1/vspherecpiconfig_types.go [hunk 1]
```diff
@@ -70,20 +70,16 @@ type NSXT struct {
...
 
-	// The name of secret that stores CPI configuration
-	// +kubebuilder:validation:Optional
-	SecretName string `json:"secretName"`
-
-	// The namespace of secret that stores CPI configuration
+	// The secret that stores CPI configuration
 	// +kubebuilder:validation:Optional
-	SecretNamespace string `json:"secretNamespace"`
+	CPISecretRef *v1.SecretReference `json:"cpiSecretRef"`
```
Comment (lubronzhan) [inline apis/cpi/v1alpha1/vspherecpiconfig_types.go:75]: What's this?

Comment (HanFa) [reply apis/cpi/v1alpha1/vspherecpiconfig_types.go]: removed. Currently only keep `CredentialRef` as the only source of NSXT credential.

File: addons/controllers/cpi/vspherecpiconfig_controller.go [hunk 2]
```diff
@@ -143,7 +147,7 @@ func (r *VSphereCPIConfigReconciler) reconcileVSphereCPIConfigNormal(ctx context
 	}
 
 	// deploy the provider service account for paravirtual mode
-	if cpiConfig.Spec.VSphereCPI.Mode == VSphereCPIParavirtualMode {
+	if *cpiConfig.Spec.VSphereCPI.Mode == VSphereCPIParavirtualMode {
```
Comment (lubronzhan) [inline addons/controllers/cpi/vspherecpiconfig_controller.go:150]: When referencing, the * is not needed

File: addons/controllers/cpi/vspherecpiconfig_utils.go [hunk 3]
```diff
@@ -208,15 +190,16 @@ func (r *VSphereCPIConfigReconciler) mapCPIConfigToDataValuesParavirtual(_ conte
 
 // mapCPIConfigToDataValues maps VSphereCPIConfig CR to data values
 func (r *VSphereCPIConfigReconciler) mapCPIConfigToDataValues(ctx context.Context, cpiConfig *cpiv1alpha1.VSphereCPIConfig, cluster *clusterapiv1beta1.Cluster) (*VSphereCPIDataValues, error) {
-	switch cpiConfig.Spec.VSphereCPI.Mode {
+	mode := *cpiConfig.Spec.VSphereCPI.Mode
```
Comment (lubronzhan) [inline addons/controllers/cpi/vspherecpiconfig_utils.go:193]: Same here, `*`  could be removed

File: addons/controllers/vspherecpiconfig_controller_test.go [hunk 4]
```diff
@@ -94,9 +94,9 @@ var _ = Describe("VSphereCPIConfig Reconciler", func() {
 				if err := k8sClient.Get(ctx, key, config); err != nil {
 					return false
 				}
-				Expect(config.Spec.VSphereCPI.Mode).Should(Equal("vsphereCPI"))
-				Expect(config.Spec.VSphereCPI.Region).Should(Equal("test-region"))
-				Expect(config.Spec.VSphereCPI.Zone).Should(Equal("test-zone"))
+				Expect(*config.Spec.VSphereCPI.Mode).Should(Equal("vsphereCPI"))
```
Comment (lubronzhan) [inline addons/controllers/vspherecpiconfig_controller_test.go:97]: '*' this could be removed

File: apis/cpi/v1alpha1/vspherecpiconfig_types.go [hunk 5]
```diff
@@ -13,25 +13,39 @@ type VSphereCPIConfigSpec struct {
...
-type NSXTRoute struct {
+type NSXTRouteConfig struct {
 	// NSX-T T0/T1 logical router path
 	// +kubebuilder:validation:Optional
-	RouterPath string `json:"routerPath"`
+	RouterPath *string `json:"routerPath"`
+}
 
-	// Cluster CIDR
+type VMNetwork struct {
```
Comment (lubronzhan) [inline apis/cpi/v1alpha1/vspherecpiconfig_types.go:22]: Nice!

File: apis/cpi/v1alpha1/vspherecpiconfig_types.go [hunk 6]
```diff
@@ -13,25 +13,39 @@ type VSphereCPIConfigSpec struct {
...
+	// +kubebuilder:validation:Optional
+	ExcludeExternalSubnetCidr *string `json:"excludeExternalSubnetCidr"`
 }
 
 type NSXT struct {
 	// A flag that enables pod routing
 	// +kubebuilder:validation:Optional
 	// +kubebuilder:default:=false
-	PodRoutingEnabled bool `json:"podRoutingEnabled"`
+	PodRoutingEnabled *bool `json:"podRoutingEnabled"`
```
Comment (lubronzhan) [inline apis/cpi/v1alpha1/vspherecpiconfig_types.go:44]: For optional one, you need omitempty in the json body as well.

File: addons/controllers/vspherecpiconfig_controller_test.go (hunk 4 above)
Comment (HanFa) [reply addons/controllers/vspherecpiconfig_controller_test.go:97]: not quite, `config.Spece.VSphereCPI.Mode` is of type `*string` now

File: apis/cpi/v1alpha1/vspherecpiconfig_types.go (hunk 6 above)
Comment (HanFa) [reply apis/cpi/v1alpha1/vspherecpiconfig_types.go:44]: ok, let me add it

File: addons/controllers/cpi/vspherecpiconfig_controller.go [hunk 7]
```diff
@@ -61,6 +61,10 @@ func (r *VSphereCPIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	if cluster == nil {
 		return ctrl.Result{}, err // no need to requeue if cluster is not found
 	}
+	if cpiConfig.Spec.VSphereCPI.Mode == nil {
+		r.Log.Info("CPI mode is not provided.")
```
Comment (lubronzhan) [inline addons/controllers/cpi/vspherecpiconfig_controller.go:65]: vsphere CPI mode

**Review by lubronzhan — APPROVED** [review]
/lgtm

**Review by danniel1205 — APPROVED** [review]
lgtm

**Pull request merged**